File: apps/frontend/src/app/types/blog.ts
export type UserRole = "admin" | "author";

export type User = {
  id: number;
  username: string;
  email: string;
  role: UserRole;
  is_active: boolean;
  created_at: string;
};

export type BlogStatus =
  | "draft"
  | "pending_review"
  | "approved"
  | "rejected"
  | "withdrawn";

export type BlogReviewComment = {
  id: number;
  status: BlogStatus;
  comment: string;
  created_at: string;
  admin: User;
};

export interface Blog {
  id: number;
  title: string;
  summary: string;
  content: string;
  tags: string[];
  status: BlogStatus;
  submitted_at: string | null;
  reviewed_at: string | null;
  created_at: string;
  updated_at: string;
  author: User;
  review_comments: BlogReviewComment[];
}

File: apps/backend/app/schemas/blog.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.user import UserRead
from app.utils.enums import BlogStatus


class BlogCreate(BaseModel):
    title: str = Field(min_length=3, max_length=255)
    summary: str = Field(min_length=10, max_length=500)
    content: str = Field(min_length=20)
    tags: list[str] = Field(default_factory=list)


class BlogUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=3, max_length=255)
    summary: str | None = Field(default=None, min_length=10, max_length=500)
    content: str | None = Field(default=None, min_length=20)
    tags: list[str] | None = None


class BlogReviewAction(BaseModel):
    status: BlogStatus
    comment: str = Field(min_length=3, max_length=1000)


class BlogReviewCommentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    status: BlogStatus
    comment: str
    created_at: datetime
    admin: UserRead


class BlogRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    summary: str
    content: str
    tags: list[str]
    status: BlogStatus
    submitted_at: datetime | None
    reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    author: UserRead
    review_comments: list[BlogReviewCommentRead] = []


class BlogListResponse(BaseModel):
    items: list[BlogRead]
